velox/dwio/parquet/reader/PageReader.cpp

```diff
@@ -75,53 +75,64 @@ void PageReader::seekToPage(int64_t row) {
   }
 }
 
+// The size of Parquet Page Header remains unknown until the Header is parsed
+// because it is in Thrift format. To read Parquet Page Header, we are using an
+// adaptive approach to guess the Page Header size.
+
+// TODO: Refactor the inputStreams to have a general interface where read() can
+// be applied on different chunks of memory.
 PageHeader PageReader::readPageHeader(int64_t remainingSize) {
   // Note that sizeof(PageHeader) may be longer than actually read
   std::shared_ptr<thrift::ThriftBufferedTransport> transport;
   std::unique_ptr<apache::thrift::protocol::TCompactProtocolT<
       thrift::ThriftBufferedTransport>>
       protocol;
-  char copy[sizeof(PageHeader)];
-  bool wasInBuffer = false;
-  if (bufferEnd_ == bufferStart_) {
-    const void* buffer;
-    int32_t size;
-    inputStream_->Next(&buffer, &size);
-    bufferStart_ = reinterpret_cast<const char*>(buffer);
-    bufferEnd_ = bufferStart_ + size;
-  }
-  if (bufferEnd_ - bufferStart_ >= sizeof(PageHeader)) {
-    wasInBuffer = true;
-    transport = std::make_shared<thrift::ThriftBufferedTransport>(
-        bufferStart_, sizeof(PageHeader));
-    protocol = std::make_unique<apache::thrift::protocol::TCompactProtocolT<
-        thrift::ThriftBufferedTransport>>(transport);
-  } else {
+
+  uint32_t pageHeaderBufferSize = kPageHeaderBufferSize_;
+  uint32_t oldPageHeaderBufferSize = 0;
+
+  BufferPtr copyBuffer =
+      AlignedBuffer::allocate<char>(pageHeaderBufferSize, &pool_);
+
+  uint32_t attempt = 0;
+  for (attempt = 0; attempt < kMaxAttempts_; attempt++) {
     dwio::common::readBytes(
-        std::min<int64_t>(remainingSize, sizeof(PageHeader)),
+        std::min<int64_t>(
+            remainingSize, pageHeaderBufferSize - oldPageHeaderBufferSize),
         inputStream_.get(),
-        &copy,
+        copyBuffer->asMutable<char>() + oldPageHeaderBufferSize,
         bufferStart_,
         bufferEnd_);
-
     transport = std::make_shared<thrift::ThriftBufferedTransport>(
-        copy, sizeof(PageHeader));
+        copyBuffer->asMutable<char>(), pageHeaderBufferSize);
     protocol = std::make_unique<apache::thrift::protocol::TCompactProtocolT<
         thrift::ThriftBufferedTransport>>(transport);
+
+    try {
+      PageHeader pageHeader;
+      // Use the Thrift generated read() function to read Parquet page header.
+      uint64_t readBytes = pageHeader.read(protocol.get());
+      pageDataStart_ = pageStart_ + readBytes;
+      std::vector<uint64_t> start = {pageDataStart_};
+      dwio::common::PositionProvider position(start);
+      inputStream_->seekToPosition(position);
+      bufferStart_ = bufferEnd_ = nullptr;
+      copyBuffer = nullptr;
+      return pageHeader;
+    } catch (std::exception const&) {
```

Comment:
Can we be more specific about the exception thrown here?

Reply:
These exceptions are from the thrift generated code. So in readPageHeader(), I added two kind is Velox errors: page too large or unable to parse.

Reply:
It's just if we can narrow down the exception type when the message is incomplete and put it here, we are much safer

Reply:
The Thrift generated code just throws `std::exception` , instead of more specific types. Let me try to narrow down the Thrift code part.